File: RTNPerformance/android/src/main/java/com/rtnperformance/PerformanceLoggerStorage.java
package com.rtnperformance;


import android.os.Environment;
import android.util.Log;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class PerformanceLoggerStorage {

    String START_TIME = "Start";
    String END_TIME = "End";

    String MEAN = "Mean";
    String STANDARD_DEVIATION = "Standard Deviation";

    String ERROR_RATE = "Error Rate";

    Double Z_VALUE = 1.97;

    String FOLDER_NAME = "D11PerformanceLog";
    static HashMap<String, HashMap<String, List<Double>>> logs =  new HashMap<>();
    static HashMap<String, HashMap<String, Double>> result = new HashMap<>();

    void addStartTime(String name, String timestamp) {
        if (logs.containsKey(name) == true) {
            HashMap<String, List<Double>> item = logs.get(name);
            List<Double> startTimeList = item.get(START_TIME);
            startTimeList.add(Double.parseDouble(timestamp));

        } else {
            HashMap<String, List<Double>> item = new HashMap<>();
            List<Double> startTimeList = new ArrayList<>();
            startTimeList.add(Double.parseDouble(timestamp));
            item.put(START_TIME, startTimeList);
            item.put(END_TIME, new ArrayList<>());
            logs.put(name, item);
        }

        writeToFile(name, START_TIME + " " + timestamp);
    }

    void addEndTime(String name, String timestamp) {
        if (logs.containsKey(name) == true) {
            HashMap<String, List<Double>> item = logs.get(name);
            List<Double> endTimeList = item.get(END_TIME);
            endTimeList.add(Double.parseDouble(timestamp));

        } else {
            HashMap<String, List<Double>> item = new HashMap<>();
            List<Double> endTimeList = new ArrayList<>();
            endTimeList.add(Double.parseDouble(timestamp));
            item.put(START_TIME, new ArrayList<>());
            item.put(END_TIME, endTimeList);
            logs.put(name, item);
        }

        writeToFile(name, END_TIME + " " + timestamp);
    }

    void writeToFile(String fileName, String contentToWrite) {
       try {
           File file = new File(Environment.getExternalStorageDirectory(), FOLDER_NAME);


           if (!file.exists()) {
               file.mkdirs();
           }

           File logFile = new File(Environment.getExternalStorageDirectory().toString() + "/" + FOLDER_NAME, fileName + ".txt");

           if (!logFile.exists()) {

               logFile.createNewFile();
           }

           BufferedWriter fileWriter = new BufferedWriter(new FileWriter(logFile, true));

           fileWriter.write(contentToWrite);
           fileWriter.newLine();
           fileWriter.close();

       } catch (Exception e) {
           Log.d(":::: Exception: ", e.toString());
       }
    }

    void resetLogs() {
        logs = new HashMap<>();
        result = new HashMap<>();
    }

    HashMap<String, HashMap<String, Double>> generateReport() {

        if (logs.size() == 0) {
            return result;
        }

        Iterator<Map.Entry<String, HashMap<String, List<Double>>>> iterator = logs.entrySet().iterator();

        while (iterator.hasNext()) {

            Map.Entry<String, HashMap<String, List<Double>>> entry = iterator.next();

            String key = entry.getKey();
            HashMap<String, List<Double>> value = entry.getValue();

            List<Double> startTime = value.get(START_TIME);
            List<Double> endTime = value.get(END_TIME);

            List<Double> difference = new ArrayList<>();

            int listSize = startTime.size();

            for (int i = 0; i < listSize; i++) {
                difference.add(Math.abs(endTime.get(i) - startTime.get(i)));
            }

            double mean = this.getAverage(difference) / 1000;
            double standardDeviation = this.getStandardDeviation(mean, difference) / 1000;
            double errorRate = this.getErrorRate(standardDeviation, difference.size()) / 1000;

            HashMap<String, Double> resultMap = new HashMap<>();

            writeToFile(key, MEAN + " " + mean);
            writeToFile(key, STANDARD_DEVIATION + " " + standardDeviation);
            writeToFile(key, ERROR_RATE + " " + errorRate);

            resultMap.put(MEAN, mean);
            resultMap.put(STANDARD_DEVIATION, standardDeviation);
            resultMap.put(ERROR_RATE, errorRate);

            result.put(key, resultMap);

        }
        return result;
    }

    Double getStandardDeviation(double mean, List<Double> timestamps) {
        List<Double> squaredDifferences = new ArrayList<>();
        int size = timestamps.size();

        for (int i = 0; i < size; i++) {
            squaredDifferences.add(Math.pow(timestamps.get(i) - mean, 2));
        }

        double squaredDifferencesSum = 0;

        for (int j = 0; j < squaredDifferences.size(); j++) {
            squaredDifferencesSum = squaredDifferencesSum + squaredDifferences.get(j);
        }

        double variance = squaredDifferencesSum / size;

        return  Math.sqrt(variance);

    }

    Double getErrorRate(double standardDeviation, int size) {
        return (standardDeviation * Z_VALUE * 100) / Math.sqrt(size);
    }

    Double getAverage(List<Double> timestamps) {

        double sum = 0;
        int size = timestamps.size();

        for (int i = 0; i < size; i++) {
            sum = sum + timestamps.get(i);
        }
        return (sum / size);
    }


    String getLogs() {
        return logs.toString();
    }

}

File: RNPerformance/ios/PerformanceLoggerStorage.swift
import Foundation

@objc
public class PerformanceLoggerStorage: NSObject {
  
   @objc
   public static let logger = PerformanceLoggerStorage()

    let START_TIME = "Start"
    let END_TIME = "End"

    let MEAN = "Mean"
    let STANDARD_DEVIATION = "Standard Deviation"
    let ERROR_RATE = "Error Rate"

    let Z_VALUE = 1.97

    let FOLDER_NAME = "D11PerformanceLog"
    static var logs: [String: [String: [Double]]] = [:]
    static var result: [String: [String: Double]] = [:]

    @objc
    public func addStartTime(name: String, timestamp: String) {
        if var item = PerformanceLoggerStorage.logs[name] {
            item[START_TIME, default: []].append(Double(timestamp)!)
            PerformanceLoggerStorage.logs[name] = item
        } else {
            var item: [String: [Double]] = [:]
            item[START_TIME] = [Double(timestamp)!]
            item[END_TIME] = []
            PerformanceLoggerStorage.logs[name] = item
        }

        writeToFile(fileName: name, contentToWrite: "\(START_TIME) \(timestamp)")
    }

    @objc
    public func addEndTime(name: String, timestamp: String) {
        if var item = PerformanceLoggerStorage.logs[name] {
            item[END_TIME, default: []].append(Double(timestamp)!)
            PerformanceLoggerStorage.logs[name] = item
        } else {
            var item: [String: [Double]] = [:]
            item[START_TIME] = []
            item[END_TIME] = [Double(timestamp)!]
            PerformanceLoggerStorage.logs[name] = item
        }

        writeToFile(fileName: name, contentToWrite: "\(END_TIME) \(timestamp)")
    }

    func writeToFile(fileName: String, contentToWrite: String) {
        do {
            let fileManager = FileManager.default
            let documentsURL = try fileManager.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
            let folderURL = documentsURL.appendingPathComponent(FOLDER_NAME)
            try fileManager.createDirectory(at: folderURL, withIntermediateDirectories: true, attributes: nil)

            let logFileURL = folderURL.appendingPathComponent("\(fileName).txt")

            if !fileManager.fileExists(atPath: logFileURL.path) {
                fileManager.createFile(atPath: logFileURL.path, contents: nil, attributes: nil)
            }

            let fileWriter = try FileHandle(forWritingTo: logFileURL)
            fileWriter.seekToEndOfFile()
            fileWriter.write(contentToWrite.data(using: .utf8)!)
            fileWriter.write("\n".data(using: .utf8)!)
            fileWriter.closeFile()
        } catch {
            print("Error writing to file: \(error)")
        }
    }

    @objc
    public func resetLogs() {
        PerformanceLoggerStorage.logs = [:]
        PerformanceLoggerStorage.result = [:]
    }

    @objc
    public func generateReport() -> [String: [String: Double]] {
        if PerformanceLoggerStorage.logs.isEmpty {
            return PerformanceLoggerStorage.result
        }

        for (key, value) in PerformanceLoggerStorage.logs {
            let startTime = value[START_TIME] ?? []
            let endTime = value[END_TIME] ?? []

            var difference: [Double] = []

            for i in 0..<startTime.count {
                difference.append(abs(endTime[i] - startTime[i]))
            }

            let mean = getAverage(timestamps: difference) / 1000.0
            let standardDeviation = getStandardDeviation(mean: mean, timestamps: difference) / 1000.0
            let errorRate = getErrorRate(standardDeviation: standardDeviation, size: difference.count) / 1000.0

            writeToFile(fileName: key, contentToWrite: "\(MEAN) \(mean)")
            writeToFile(fileName: key, contentToWrite: "\(STANDARD_DEVIATION) \(standardDeviation)")
            writeToFile(fileName: key, contentToWrite: "\(ERROR_RATE) \(errorRate)")

            PerformanceLoggerStorage.result[key] = [
                MEAN: mean,
                STANDARD_DEVIATION: standardDeviation,
                ERROR_RATE: errorRate
            ]
        }

        return PerformanceLoggerStorage.result
    }

    public func getStandardDeviation(mean: Double, timestamps: [Double]) -> Double {
        let squaredDifferences = timestamps.map { pow($0 - mean, 2) }
        let squaredDifferencesSum = squaredDifferences.reduce(0, +)
        let variance = squaredDifferencesSum / Double(timestamps.count)
        return sqrt(variance)
    }

    func getErrorRate(standardDeviation: Double, size: Int) -> Double {
        return (standardDeviation * Z_VALUE * 100) / sqrt(Double(size))
    }

    func getAverage(timestamps: [Double]) -> Double {
        let sum = timestamps.reduce(0, +)
        return sum / Double(timestamps.count)
    }

    @objc
    public func getLogs() -> String {
        return PerformanceLoggerStorage.logs.description
    }
}
